File: java_version/mixed_recall/token_replace/token_replace.java
package mixed_recall.token_replace;

import java.util.Map;
// token replace
// Write a function, tokenReplace, that takes in an unordered_map of tokens
//  and a string. The function should return a new string where tokens are replaced.

// Tokens are enclosed in a pair of "$". You can assume that the input string
// is properly formatted. Tokens should be replaced from left to right in the string (see test_05).
public class token_replace {

    public static String tokenReplace(String s, Map<String, String> tokens) {
        return "";
    }

    public static void main(String[] args) {

        Map<String, String> tokens = Map.of(

                "$LOCATION$", "park",
                "$ANIMAL$", "dog");
        System.out.println(tokenReplace("Walk the $ANIMAL$ in the $LOCATION$!", tokens));
        // -> 'Walk the dog in the park!'
        tokens = Map.of(
                "$ADJECTIVE$", "quick",
                "$VERB$", "hopped",
                "$DIRECTION$", "North");
        System.out.println(tokenReplace("the $ADJECTIVE$ fox $VERB$ $ADJECTIVE$ly $DIRECTION$ward", tokens));
        // -> 'the quick fox hopped quickly Northward'
        tokens = Map.of(
                "$greeting$", "hey programmer");
        System.out.println(tokenReplace("his greeting is always $greeting$.", tokens));
        // -> 'his greeting is always hey programmer.'
        tokens = Map.of(
                "$A$", "lions",
                "$B$", "tigers",
                "$C$", "bears");
        System.out.println(tokenReplace("$A$$B$$C$, oh my.", tokens));
        // -> 'lionstigersbears, oh my.'
        tokens = Map.of(
                "$A$", "lions",
                "$B$", "tigers",
                "$C$", "bears");
        System.out.println(tokenReplace("$B$", tokens));
        // -> 'tigers'
        tokens = Map.of(
                "$second$", "beta",
                "$first$", "alpha",
                "$third$", "gamma");
        System.out.println(tokenReplace("$first$second$third$", tokens));
        // -> 'alphasecondgamma'

    }
}

File: javascript_version/mixed_recall/token_replace/token_replace.js
// token replace
// Write a function, tokenReplace, that takes in an unordered_map of tokens
//  and a string. The function should return a new string where tokens are replaced.

// Tokens are enclosed in a pair of "$". You can assume that the input string 
// is properly formatted. Tokens should be replaced from left to right in the string (see test_05)

// two pointers method iterative  time and spacce both O(n)
const tokenReplace = (s, tokens) => {
    // todo
    if (s.length === 0) { return s; }
    let i = 0;
    let j = 1;
    let output = "";
    while (i < s.length) {
        if (s[i] !== "$") {
            output += s[i];
            i++;
            j = i + 1;
        }
        else if (s[j] !== "$") {
            j++;
        }
        else {
            let key = s.slice(i, j + 1);
            output += tokens[key];
            i = j + 1;
            j = i + 1;
        }
    }
    return output;

};



let tokens = {
    '$LOCATION$': 'park',
    '$ANIMAL$': 'dog',
};
console.log(tokenReplace('Walk the $ANIMAL$ in the $LOCATION$!', tokens));
// -> 'Walk the dog in the park!'
tokens = {
    '$ADJECTIVE$': 'quick',
    '$VERB$': 'hopped',
    '$DIRECTION$': 'North'
};
console.log(tokenReplace('the $ADJECTIVE$ fox $VERB$ $ADJECTIVE$ly $DIRECTION$ward', tokens));
// -> 'the quick fox hopped quickly Northward'
tokens = {
    '$greeting$': 'hey programmer',
};
console.log(tokenReplace('his greeting is always $greeting$.', tokens));
// -> 'his greeting is always hey programmer.'
tokens = {
    '$A$': 'lions',
    '$B$': 'tigers',
    '$C$': 'bears',
};
console.log(tokenReplace('$A$$B$$C$, oh my.', tokens));
// -> 'lionstigersbears, oh my.'
tokens = {
    '$A$': 'lions',
    '$B$': 'tigers',
    '$C$': 'bears',
};
console.log(tokenReplace('$B$', tokens));
// -> 'tigers'
tokens = {
    '$second$': 'beta',
    '$first$': 'alpha',
    '$third$': 'gamma',
};
console.log(tokenReplace('$first$second$third$', tokens));
// -> 'alphasecondgamma'